src/build_vehicle_gui.cpp

```diff
@@ -1010,6 +1304,67 @@ void DisplayVehicleSortDropDown(Window *w, VehicleType vehicle_type, int selecte
 	ShowDropDownMenu(w, _engine_sort_listing[vehicle_type], selected, button, 0, hidden_mask);
 }
 
+bool IsEnginePassFilter(const VehicleFilter& filter, const Engine* engine) {
+	if (!filter.show_hidden_engines && engine->IsHidden(_local_company)) return false;
+	if (!IsEngineBuildable(engine->index, VEH_TRAIN, _local_company)) return false;
+
+	const RailFilter& rail_filter = filter.rail;
+	uint hasPower = engine->GetPower() > 0;
+	bool hasCapacity = engine->CanCarryCargo();
+
+	bool is_locomotive = !hasCapacity && hasPower;
+	bool is_motor_wagon = hasCapacity && hasPower;
+	bool is_wagon = hasCapacity && !hasPower;
+
+	if (!is_locomotive && !is_motor_wagon && !is_wagon) {
+		return false;
+	}
+	if (is_locomotive && !rail_filter.locomotive) {
+		return false;
+	}
+	if (is_motor_wagon && !rail_filter.motor_wagon) {
+		return false;
+	}
+	if (is_wagon && !rail_filter.wagon) {
+		return false;
+	}
+	if (!is_locomotive && !CargoFilterFunc(&engine->index, filter.cargo.type)) {
+		return false;
+	}
+	if (is_wagon) {
+		// speed in km/h
+		uint speed = engine->GetDisplayMaxSpeed();
+
+		if (rail_filter.train.speed > speed) return false;
+	}
+	if (is_locomotive && rail_filter.train.weight > 0 && rail_filter.train.speed > 0) {
+		// speed in km/h
+		uint speed = engine->GetDisplayMaxSpeed();
+		// power in kW
+		uint power = engine->GetPower() * 1000 / 735;
```

Comment:
Magic numbers... I think this calculation needs to be extracted into its own function, or otherwise be explained.